### ISSUE TITLE
Incorrect handling of midnight when specified as "12:00 am", and of invalid times

### ISSUE THREAD

**Issue by smangelsdorf** [taxjar/date_time_parser#42]
(Using `date_time_parser` 1.1.1)

Noticed this while parsing some date+time strings received from a real service. In particular the 12:00am case is not handled the way I would expect.

```
iex(1)> date = "2021-03-27"
"2021-03-27"
iex(2)> DateTimeParser.parse!("#{date} 12:00 am")
~N[2021-03-27 12:00:00]
iex(3)> DateTimeParser.parse!("#{date} 12:00 pm")
~N[2021-03-27 12:00:00]
iex(4)> DateTimeParser.parse!("#{date} 00:00 am")
~N[2021-03-27 00:00:00]
```

Also noticed that "00:00 pm" is (incorrectly?) parsed as valid:

```
iex(5)> DateTimeParser.parse!("#{date} 00:00 pm")
~N[2021-03-27 12:00:00]
```

Doing some more experimentation I also found that some other invalid times are parsed as valid:

```
iex(7)> DateTimeParser.parse!("#{date} 25:00 am")
~N[2021-03-27 02:05:00]
iex(8)> DateTimeParser.parse!("#{date} 41:00 am")
~N[2021-03-27 04:01:00]
iex(9)> DateTimeParser.parse!("#{date} 9000:00 am")
~N[2021-03-27 09:00:00]
```

**Comment by dbernheisel** [comment]
Good catch. These definitely look like issues. PRs welcome. I'll also take a look and see where we can improve detection.

**Comment by dbernheisel** [comment]
The 12am issue has been fixed in 1.1.2.

For the other invalid times that are parsed, I'll need to spend more time in the tokenizer to somehow capture and reject invalid times; right now unknown tokens are ignored and I believe that is the reason for the invalid times parsed.

**Comment by dbernheisel** [comment]
I'm going to resolve this issue and open a new issue to focus on improving time parsing.
